Cannot use the `component` helper with `undefined`

```typescript
import Component from '@glint/environment-ember-loose/glimmer-component';

export default class Foo extends Component {
  public declare readonly c?: string;
}
```

```handlebars
{{#let (component this.c) as |Component|}}
  <Component />
{{/let}}
```

```
app/components/foo.hbs:1:19 - error TS2769: No overload matches this call.
  Overload 1 of 3, '(args: EmptyObject, component: "sum"): typeof Sum', gave the following error.
    Argument of type 'string | undefined' is not assignable to parameter of type '"sum"'.
      Type 'undefined' is not assignable to type '"sum"'.
  Overload 2 of 3, '(args: {}, component: "sum"): new () => Invokable<(args: Omit<EmptyObject, never> & Partial<Pick<EmptyObject, never>>) => number>', gave the following error.
    Argument of type 'string | undefined' is not assignable to parameter of type '"sum"'.
      Type 'undefined' is not assignable to type '"sum"'.
  Overload 3 of 3, '(args: {}, component: new (...args: unknown[]) => Invokable<(args: unknown) => AcceptsBlocks<any, any>>): new () => Invokable<(args: Omit<unknown, never> & Partial<Pick<unknown, never>>) => AcceptsBlocks<...>>', gave the following error.
    Argument of type 'string | undefined' is not assignable to parameter of type 'new (...args: unknown[]) => Invokable<(args: unknown) => AcceptsBlocks<any, any>>'.
      Type 'undefined' is not assignable to type 'new (...args: unknown[]) => Invokable<(args: unknown) => AcceptsBlocks<any, any>>'.

1 {{#let (component this.c) as |Component|}}
                    ~~~~~~
```

There is also another error on the second line:

```
app/components/foo.hbs:2:3 - error TS2345: Argument of type 'number' is not assignable to parameter of type 'AcceptsBlocks<any, any>'.

2   <Component />
    ~~~~~~~~~~~~~
```

But I guess that's because of the first one?

P.S. I had a `sum` helper that I needed for something else, I see that it shows here in the error message. But I don't think it matters.

--- a/packages/environment-ember-loose/-private/intrinsics/component.d.ts
+++ b/packages/environment-ember-loose/-private/intrinsics/component.d.ts
@@ -26,6 +26,9 @@ type PartiallyAppliedComponent<AllArgs, GivenArgs, Return> = Invokable<
 export type ComponentKeyword<Registry> = DirectInvokable<{
   // {{component "some-name"}}
   <Name extends keyof Registry>(args: EmptyObject, component: Name): Registry[Name];
+  <Name extends keyof Registry>(args: EmptyObject, component: Name | null | undefined):
+    | Registry[Name]
+    | null;
 
   // {{component "some-name" arg=value}}
   <Name extends keyof Registry, GivenArgs extends Partial<RegistryComponentArgs<Registry, Name>>>(
@@ -36,6 +39,16 @@ export type ComponentKeyword<Registry> = DirectInvokable<{
     GivenArgs,
     RegistryComponentReturn<Registry, Name>
   >;
+  <Name extends keyof Registry, GivenArgs extends Partial<RegistryComponentArgs<Registry, Name>>>(
+    args: GivenArgs,
+    component: Name | null | undefined
+  ):
+    | null
+    | (new () => PartiallyAppliedComponent<
+        RegistryComponentArgs<Registry, Name>,
+        GivenArgs,
+        RegistryComponentReturn<Registry, Name>
+      >);
 
   // {{component someCurriedComponent arg=value}}
   <
@@ -47,4 +60,16 @@ export type ComponentKeyword<Registry> = DirectInvokable<{
     args: GivenArgs,
     component: new (...args: ConstructorArgs) => Invokable<(args: Args) => Return>
   ): new () => PartiallyAppliedComponent<Args, GivenArgs, Return>;
+  <
+    Args,
+    GivenArgs extends Partial<Args>,
+    Return extends AcceptsBlocks<any, any>,
+    ConstructorArgs extends unknown[]
+  >(
+    args: GivenArgs,
+    component:
+      | (new (...args: ConstructorArgs) => Invokable<(args: Args) => Return>)
+      | null
+      | undefined
+  ): null | (new () => PartiallyAppliedComponent<Args, GivenArgs, Return>);
 }>;

--- a/packages/environment-ember-loose/-private/intrinsics/component.d.ts
+++ b/packages/environment-ember-loose/-private/intrinsics/component.d.ts
@@ -17,6 +17,12 @@ type RegistryComponentReturn<Registry, T extends keyof Registry> = Registry[T] e
   ? Return
   : unknown;
 
+type PartiallyAppliedComponent<AllArgs, GivenArgs, Return> = Invokable<
+  (
+    args: Omit<AllArgs, keyof GivenArgs> & Partial<Pick<AllArgs, keyof GivenArgs & keyof AllArgs>>
+  ) => Return
+>;
+
 export type ComponentKeyword<Registry> = DirectInvokable<{
   // {{component "some-name"}}
   <Name extends keyof Registry>(args: EmptyObject, component: Name): Registry[Name];
@@ -25,17 +31,11 @@ export type ComponentKeyword<Registry> = DirectInvokable<{
   <Name extends keyof Registry, GivenArgs extends Partial<RegistryComponentArgs<Registry, Name>>>(
     args: GivenArgs,
     component: Name
-  ): new () => Invokable<{
-    (
-      args: Omit<RegistryComponentArgs<Registry, Name>, keyof GivenArgs> &
-        Partial<
-          Pick<
-            RegistryComponentArgs<Registry, Name>,
-            keyof GivenArgs & keyof RegistryComponentArgs<Registry, Name>
-          >
-        >
-    ): RegistryComponentReturn<Registry, Name>;
-  }>;
+  ): new () => PartiallyAppliedComponent<
+    RegistryComponentArgs<Registry, Name>,
+    GivenArgs,
+    RegistryComponentReturn<Registry, Name>
+  >;
 
   // {{component someCurriedComponent arg=value}}
   <
@@ -46,7 +46,5 @@ export type ComponentKeyword<Registry> = DirectInvokable<{
   >(
     args: GivenArgs,
     component: new (...args: ConstructorArgs) => Invokable<(args: Args) => Return>
-  ): new () => Invokable<{
-    (args: Omit<Args, keyof GivenArgs> & Partial<Pick<Args, keyof GivenArgs & keyof Args>>): Return;
-  }>;
+  ): new () => PartiallyAppliedComponent<Args, GivenArgs, Return>;
 }>;

--- a/packages/environment-ember-loose/__tests__/intrinsics/component.test.ts
+++ b/packages/environment-ember-loose/__tests__/intrinsics/component.test.ts
@@ -10,6 +10,8 @@ import { ComponentWithBoundArgs, ComponentLike } from '@glint/environment-ember-
 
 const componentKeyword = resolve({} as ComponentKeyword<LocalRegistry>);
 
+declare function maybe<T>(arg: T): T | undefined;
+
 type LocalRegistry = {
   string: typeof StringComponent;
   parametric: typeof ParametricComponent;
@@ -24,6 +26,9 @@ class StringComponent extends Component<{
 const NoopCurriedStringComponent = componentKeyword({}, 'string');
 const ValueCurriedStringComponent = componentKeyword({ value: 'hello' }, 'string');
 
+const MaybeNoopCurriedStringComponent = componentKeyword({}, maybe('string'));
+const MaybeValueCurriedStringComponent = componentKeyword({ value: 'hello' }, maybe('string'));
+
 // Once value is curried, it should be optional in args
 expectTypeOf(ValueCurriedStringComponent).toEqualTypeOf<
   ComponentLike<{
@@ -31,6 +36,12 @@ expectTypeOf(ValueCurriedStringComponent).toEqualTypeOf<
     Args: { value?: string };
     Yields: { default?: [string] };
   }>
+>();
+
+// {{component maybeAString}} returns Component | null
+expectTypeOf(MaybeNoopCurriedStringComponent).toEqualTypeOf<null | typeof StringComponent>();
+expectTypeOf(MaybeValueCurriedStringComponent).toEqualTypeOf<
+  null | typeof ValueCurriedStringComponent
 >();
 
 // This is also equivalent to this `ComponentWithBoundArgs` shorthand:
@@ -220,6 +231,15 @@ const DoubleCurriedComponent = componentKeyword(
   { optional: 'hi' },
   RequiredValueCurriedParametricComponent
 );
+
+const MaybeDoubleCurriedParametricComponent = componentKeyword(
+  { optional: 'hi' },
+  maybe(RequiredValueCurriedParametricComponent)
+);
+
+expectTypeOf(MaybeDoubleCurriedParametricComponent).toEqualTypeOf<
+  null | typeof DoubleCurriedComponent
+>();
 
 // Invoking the component with no args
 {

